extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -79,16 +87,27 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     @Override
     public void initialize() {
         GatewayConfig configuration = getConfigAs(GatewayConfig.class);
-        if (configuration.host == null || configuration.host.isEmpty()) {
+        if (Strings.isNullOrEmpty(configuration.host)) {
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
                     "Host must be specified in the configuration!");
             return;
         }
-        if (configuration.code == null || configuration.code.isEmpty()) {
+        if (Strings.isNullOrEmpty(configuration.code)) {
+            if(Strings.isNullOrEmpty(configuration.username) || Strings.isNullOrEmpty(configuration.userkey)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "Security code must be provided in the configuration when initializing for first time!");
+                return;
+            }
+        } else {
+            configuration = getUsernameAndKey(configuration);
```

Comment:
You are always overwriting the configuration here? What if the user has already manually configured a username/key?

Reply:
And IF the binding does changes to the config, it must use editConfiguration and updateConfiguration methods.

Reply:
If the security code is provided - then yes - the username and pre-shared key are overwritten no matter if they exists or not

Reply:
There is even a new pair created upon every new startup. Reading the IKEA docs, this is definitely not the way it is supposed to be used. Note that they ask to not permanently store the security code at all. So after having received a PSK, we should rather remove the security code from the config and keep using only the PSK from that moment on.

Reply:
This is exactly what's happening:

thing.getConfiguration().put(GatewayConfig.USERNAME, user);
thing.getConfiguration().put(GatewayConfig.USERKEY, preSharedPsk);
thing.getConfiguration().put(GatewayConfig.CODE, null);

New pair will be created ONLY IF the code is present:
if (Strings.isNullOrEmpty(configuration.code)) {
   ....
} else {
  configuration = getUsernameAndKey(configuration);
}

If not - the username and key from the configuration will be used